chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,943 @@
+//! Logic for resharding flat storage in parallel to chain processing.
+//!
+//! See [FlatStorageResharder] for more details about how the resharding takes place.
+
+use std::sync::{Arc, Mutex};
+
+use crossbeam_channel::{Receiver, Sender};
+use near_chain_configs::ReshardingHandle;
+use near_chain_primitives::Error;
+
+use tracing::{debug, error, info, warn};
+
+use crate::resharding::event_type::{ReshardingEventType, ReshardingSplitShardParams};
+use crate::types::RuntimeAdapter;
+use near_primitives::shard_layout::{account_id_to_shard_id, ShardLayout};
+use near_primitives::state::FlatStateValue;
+use near_primitives::trie_key::col::{self, ALL_COLUMNS_WITH_NAMES};
+use near_primitives::trie_key::trie_key_parsers::{
+    parse_account_id_from_access_key_key, parse_account_id_from_account_key,
+    parse_account_id_from_contract_code_key, parse_account_id_from_contract_data_key,
+    parse_account_id_from_received_data_key, parse_account_id_from_trie_key_with_separator,
+};
+use near_primitives::types::AccountId;
+use near_store::adapter::flat_store::{FlatStoreAdapter, FlatStoreUpdateAdapter};
+use near_store::adapter::StoreAdapter;
+use near_store::flat::{
+    BlockInfo, FlatStorageReadyStatus, FlatStorageReshardingStatus, FlatStorageStatus,
+    SplittingParentStatus,
+};
+use near_store::{ShardUId, StorageError};
+
+/// `FlatStorageResharder` takes care of updating flat storage when a resharding event
+/// happens.
+///
+/// On an high level, the events supported are:
+/// - #### Shard splitting
+///     Parent shard must be split into two children. The entire operation freezes the flat storage
+///     for the involved shards.
+///     Children shards are created empty and the key-values of the parent will be copied into one of them,
+///     in the background.
+///
+///     After the copy is finished the children shard will have the correct state at some past block height.
+///     It'll be necessary to perform catchup before the flat storage can be put again in Ready state.
+///     The parent shard storage is not needed anymore and can be removed.
+///
+/// The resharder has also the following properties:
+/// - Background processing: the bulk of resharding is done in a separate task, see [FlatStorageResharderScheduler]
+/// - Interruptible: a reshard operation can be interrupted through a [FlatStorageResharderController].
+///     - In the case of event `Split` the state of flat storage will go back to what it was previously.
+#[derive(Clone)]
+pub struct FlatStorageResharder {
+    runtime: Arc<dyn RuntimeAdapter>,
+    resharding_event: Arc<Mutex<Option<FlatStorageReshardingEventStatus>>>,
+}
+
+impl FlatStorageResharder {
+    /// Creates a new `FlatStorageResharder`.
+    pub fn new(runtime: Arc<dyn RuntimeAdapter>) -> Self {
+        let resharding_event = Arc::new(Mutex::new(None));
+        Self { runtime, resharding_event }
+    }
+
+    /// Resumes a resharding event that was in progress.
```

Comment:
nit: ... that was interrupted 

It feels strange to be resuming something that is in progress. 

Can you also add what can cause resharding to be halted and resumed? Is node restart / crashing the only event that may cause this?

Reply:
Answering my own question - I think a fork may also cause interruption.

Reply:
I made more clear the distinction between cancelling (to be used in the case forks) and interrupting (when the node crashes).